Expression visitors : first implementation

flomnes opened the pull request:
TODO
- [x] Add missing object types
- [x] Add missing visitors
- [x] Re-organize source files / split header/source, use the usual hierarchy with `include/XXX/`, define visitor / node types separately, etc.
- [x] Deep copy for nodes (copy operator/constructor is currently deleted because of `std::unique_ptr`)
- [x] `private` children for nodes ?
- [x] Use a generic type to avoid defining ~14 object types manually ?
- [x] Sonar issues
- [x] Separate tests (move existing calls to test, add more tests)

~~The following causes a double free because of two `std::unique_ptr` holding the same memory address~~
```cpp
Parameter* p = new Parameter(1.);
Add(p, p);
```

flomnes committed:
Expression visitors

flomnes committed:
Forgotten

flomnes committed:
Apply clang-format

flomnes committed:
Add Clone visitor

flomnes committed:
Remove a few dynamic_cast

flomnes committed:
Use std::unique_ptr for root of cloned tree

a-zakir committed:
split src files

a-zakir committed:
delete

a-zakir committed:
add header

a-zakir committed:
add .cpp files

a-zakir committed:
fix && add missing includes

flomnes committed:
Add missing header, fix build

a-zakir committed:
gathering all nodes header in one

a-zakir committed:
Merge remote-tracking branch 'origin/feature/expression-visitors' into feature/expression-visitors

flomnes committed:
Add `visit` helper

a-zakir committed:
create namespace Antares::Solver::Expressions

flomnes committed:
Use `std::shared_ptr` for memory management (#2293)

Co-authored-by: Abdoulbari Zakir <abdoulbari.zakir@gmail.com>

a-zakir committed:
test fix windows

a-zakir committed:
update

a-zakir committed:
replace array to vector

flomnes committed:
Rename visit -> dispatch

flomnes committed:
Use std::array

flomnes committed:
Clone wars

a-zakir committed:
Add pragma once [skip ci]

flomnes committed:
Use std::unique_ptr (#2298)

TODO : use `std::unique_ptr` in node constructors ?

flomnes committed:
Use memory manager instead of std::unique_ptr (#2307)

Co-authored-by: Abdoulbari Zaher <32519851+a-zakir@users.noreply.github.com>
Co-authored-by: Abdoulbari Zakir <abdoulbari.zakir@gmail.com>

a-zakir committed:
Completing Nodes List (#2303)

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes committed:
Add tests for expression visitors (#2314)

- Add Boost tests
- Remove main.cpp & "visitor" executable

---------

Co-authored-by: Abdoulbari Zakir <abdoulbari.zakir@gmail.com>

flomnes committed:
Add VariableNode properly (#2319)

```cpp
using LiteralNode = Leaf<double>;
```
This causes problems because we need distinct types, which we obviously
don't get (only aliases to the same type).

pet-mit commented on src/solver/expressions/include/antares/solver/expressions/visitors/Visitor.h:
```diff
@@ -0,0 +1,82 @@
+{
+
+template<class R>
+class Visitor
```
If we want to neatly implement the Visitor pattern I would suggest:
- Move Visitor class to the "nodes" folder, rename it to NodeVisitor
- Add a Node::accept(NodeVisitor) virtual method, and implement it in all nodes in the same way: nodeVisitor.visit(this)
- In the dispatch method, call node.accept(this) instead (and maybe find a better name, like others suggested)

(maybe we already discussed it, but I guess I wasn't able to follow before reading all the dev)

flomnes replied:
@pet-mit 
```suggestion
class NodeVisitor
```

flomnes commented on src/solver/expressions/include/antares/solver/expressions/visitors/EvalVisitor.h:
```diff
@@ -0,0 +1,41 @@
+    double visit(const ParameterNode& param) override;
+    double visit(const LiteralNode& lit) override;
+    double visit(const PortFieldNode& port_field_node) override;
+    double visit(const ComponentVariableNode& component_variable_node);
+    double visit(const ComponentParameterNode& component_parameter_node);
```
```suggestion
    double visit(const ComponentVariableNode& component_variable_node) override;
    double visit(const ComponentParameterNode& component_parameter_node) override;
```

flomnes commented on src/solver/expressions/include/antares/solver/expressions/visitors/Visitor.h:
@pet-mit Move to nodes

flomnes reviewed (commented):
+ namespaces `Node`
+ namespace `Visitor`

`Node::Add`

flomnes committed:
Add context for evaluation, VariableNode (#2317)

- Add evaluation context for variable & parameter nodes
- Add forgotten boilerplate code for variable nodes (Visitor.h)
- Add 2 simple tests: one for a variable node, one for a parameter node

---------

Co-authored-by: Abdoulbari Zakir <abdoulbari.zakir@gmail.com>

a-zakir committed:
Update src/solver/expressions/include/antares/solver/expressions/visitors/EvalVisitor.h

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes committed:
Shorter lock for mutex (#2324)

Follow the general guideline: lock resources just when we need them.
This should avoid unnecessary wait times in multi-threaded uses.

guilpier-code commented on src/tests/src/solver/expressions/test_expressions.cpp:
```diff
@@ -0,0 +1,178 @@
+
+BOOST_AUTO_TEST_CASE(print_single_literal)
+{
+    LiteralNode lit(21);
```
Instead of `LiteralNode lit(21);`, it may not be more costly to have : 
- `LiteralNode literal(21);`
- or `LiteralNode twentyOne(21);`

and it's more clear.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/Leaf.h:
```diff
@@ -0,0 +1,49 @@
+{
+// constraint T to arithmetic types?
+template<class T>
+class Leaf: public Node
```
_Draft remark_ : 
class **Leaf** : 
**Leaf** might be too vague. 
What do leaves represent eventually ?

a-zakir replied:
a node without child

guilpier-code replied:
Shouldn't we rename it into **ValueNode** (as it contains an actual value, unlike other nodes) ?

a-zakir replied:
Parameter, constant etc...

a-zakir replied:
I think that Florian has chosen the word "leaf" because it is part of the vocabulary of trees (Ast),

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/VariableNode.h:
```diff
@@ -0,0 +1,14 @@
+
+namespace Antares::Solver::Expressions
+{
+class VariableNode: public Leaf<std::string>
```
_Draft remark_ : 

class **VariableNode** is empty. Meaning this class is very similar to its parent class.
What is it for ?
The type may be important : visitor can execute a task different than it does for the parent.

a-zakir replied:
yes type is important for visitors

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/AddNode.h:
```diff
@@ -0,0 +1,40 @@
+    AddNode(Node* n1, Node* n2);
+
+    // private:
+    Node* n1_;
```
_Draft remark_ : 

```C++
    Node* n1_;
    Node* n2_;
```
Why aren't these data members **private** ?

a-zakir replied:
yes, they are private now

a-zakir committed:
Feature/add complex nodes (#2318)

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/visitors/Visitor.h:
```diff
@@ -0,0 +1,84 @@
+{
+private:
+    template<class T>
+    std::optional<R> tryType(const Node& node)
```
_Draft remark_ : 

About **tryType(...)** and its uses / calls : strange (maybe justified, to be looked at more deeply).
 
We deviate from the classic **visitor** design pattern, why do we do that ?
For what benefit ?
Is it because of the type returned by each **accept(...)** function depending on the visitor type ? 
Do we try to save code lines by **not** defining an **accept(...)** function for each node class ?
By defining this **tryType** function, it seems we the intention is to replace the polymorphism at work in the classic **visitor** form by using templates and `dynamic_cast ` operations.
What consequences of having our own visitor form ?
- Is it risky ? Unlike classic form, not really proven / experienced.
- Is it unclear ? Unexpected form of **visitor** is surprising.

guilpier-code replied:
Complications come up when **accept(...)** functions have to return a different type depending on the **visitor** : **std::string** for **PrintVisitor**, double/int for the evaluate visitor, ...
This makes the visitor classic form not adapted to this particular situation.

a-zakir committed:
add () for PrintVisitor

a-zakir committed:
update

a-zakir committed:
sonar

a-zakir committed:
sonar

a-zakir committed:
rename Visitor --> NodeVisitor

a-zakir committed:
move NodeVisitor.h

a-zakir committed:
improvements

a-zakir committed:
fix

a-zakir committed:
add LinearVisitor

a-zakir committed:
add dep-variable

a-zakir committed:
update

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,96 @@
+
+    R dispatch(const Node& node)
+    {
+        using Function = std::optional<R> (Antares::Solver::Expressions::NodeVisitor<R>::*)(
```
About `Antares::Solver::Expressions::NodeVisitor<R>::*` : 
no need to recall the **namespace** context, we're already in.
`NodeVisitor<R>::*` is enough.

Same thing for` Antares::Solver::Expressions::Node`.

a-zakir replied:
done

flomnes replied:
Nope, MSVC won't compile without full namespace qualifiers.

a-zakir committed:
revert 3 last commits

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,96 @@
+    R dispatch(const Node& node)
+    {
+        using Function = std::optional<R> (Antares::Solver::Expressions::NodeVisitor<R>::*)(
+          const Antares::Solver::Expressions::Node&);
```
Why not using **std::function** instead (more clear) ?
```C++
using Function = std::function<std::optional<R>(const Node&)>;
```

flomnes replied:
Because these are pointer to member functions. These cannot be assigned into an `std::function`, see [here](https://stackoverflow.com/questions/7582546/using-generic-stdfunction-objects-with-member-functions-in-one-class) for more details.

flomnes replied:
We could use [std::mem_fn](https://en.cppreference.com/w/cpp/utility/functional/mem_fn) for this, but the return type is unspecified, so we can't easily store a collection of it.

a-zakir replied:
you example is not a pointer to member function;

a-zakir committed:
Pr remarks

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,96 @@
+    {
+        using Function = std::optional<R> (Antares::Solver::Expressions::NodeVisitor<R>::*)(
+          const Antares::Solver::Expressions::Node&);
+        static const std::array<Function, 16> tryFunctions{
```
About `static const std::array<Function, 16> tryFunctions` : 
- This list of function objects does not have to be defined inside **dispatch** method, does it ? It bloats the **dispatch** function. Where else could we define it ? Why not outside the class itself (but in the same header) ? Or as a private data member. 
- Does it have to be **static** ? 
- Does it have to be a **std::array** ? A **std::vector** could be a tiny bit easier to define, as we don't have to give its size (unlike **std::array**). And in case of change, the size does not have to be updated (ok, not a big improvement...).
- This list contains **tryType** functions. 
  About these functions : technically, they do 2 things : 
  - they turn the argument's polymorphic type into the expected type (template parameter) 
  - they run **visit** function on the object of the expected type (if it is of the expected type)
  
  So, to me, name **tryType** is vague, and does not describe what the function actually does.
  
About this last point, we could instead : 
Redefine **tryType** template function like this : 
```C++
template<class T>
T* tryType(const Node& node)
{
    return dynamic_cast<const T*>(&node);
}
```
and change a bit the **dispatch** function : 
```c++
R dispatch(const Node& node)
{
    ...
    for (auto f: tryTypes) // tryFunctions renamed into tryTypes
    {
        if (auto* actualNode = f(node))
        {
            return visit(*actualNode);
        }
    }

    logs.error() << "Antares::Solver::Expressions Visitor: unsupported Node!";
    return R();
}
```
It's more simple like this : 
- the **std::optional<R>** disappeared.
- template **tryType** is shorter and is named according to what it actually does

a-zakir replied:
@flomnes

flomnes replied:
> About `static const std::array<Function, 16> tryFunctions` :
> 
>     * This list of function objects does not have to be defined inside **dispatch** method, does it ? It bloats the **dispatch** function. Where else could we define it ? Why not outside the class itself (but in the same header) ? Or as a private data member.
> 
Could be outside
>     * Does it have to be **static** ?
>
Yes, we don't want to rebuild it at every function call (1000s of function calls) for obvious reasons 
>     * Does it have to be a **std::array** ? A **std::vector** could be a tiny bit easier to define, as we don't have to give its size (unlike **std::array**). And in case of change, the size does not have to be updated (ok, not a big improvement...).
> 
Yes it can be a `std::vector`
>     * This list contains **tryType** functions.
>       About these functions : technically, they do 2 things :
>       
>       * they turn the argument's polymorphic type into the expected type (template parameter)
>       * they run **visit** function on the object of the expected type (if it is of the expected type)
>       
>       So, to me, name **tryType** is vague, and does describe what the function actually does.
> 
> 
> About this last point, we could instead : Redefine **tryType** template function like this :
> 
> ```c++
> template<class T>
> T* tryType(const Node& node)
> {
>     return dynamic_cast<const T*>(&node);
> }
> ```
> 
> and change a bit the **dispatch** function :
> 
> ```c++
> R dispatch(const Node& node)
> {
>     ...
>     for (auto f: tryTypes) // tryFunctions renamed into tryTypes
>     {
>         if (auto* actualNode = f(node))
>         {
>             return visit(*actualNode);
>         }
>     }
> 
>     logs.error() << "Antares::Solver::Expressions Visitor: unsupported Node!";
>     return R();
> }
> ```
> 
> It's more simple like this :
> 
>     * the **std::optional** disappeared.
> 
>     * template **tryType** is shorter and is named according to what it actually does

I agree that your code is simpler. But it doesn't compile because you can't (easily at least) iterate over types in C++.

Specifically, your code doesn't compile because at some point you want to store a collection of functions (just before the `for` loop), but these functions don't have the same signature. So we can't make a collection of them.

guilpier-code replied:
ok, then I close this discussion.

a-zakir committed:
fix

a-zakir committed:
namespace conflicts

a-zakir committed:
Pr remark

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/BinaryNode.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace Antares::Solver::Nodes
+{
+class BinaryNode: public ParentNode<2>
```
We should make the nodes class hierarchy the most shallow as possible. Make a deep class hierarchy is considered as a bad practice, as you know, and we should normally favor composition over inheritance.
That said, it seems like classes **BinaryNode** and **UnaryNode** are useless.
Indeed : 
- As you know, they add very little behavior (only a constructor)
- we have already a template class **ParentNode\<n\>**

If it turns out that we really need **ParentNode\<n\>**, we can specialize its constructor, depending on the value of **N**.
Furthermore, **ParentNode** is not expressive : it means anything you want. 
**MultipleNode** would be a better name (for instance).
 
About class **ParentNode\<n\>** : 
This class is used to have the **[] operator** common to its children classes. 
For now, we use it for **n = 1** and **n = 2**.
- **n = 1** : using a **[] operator** is overkill. So any **Node** class that contains another **Node** should not inherit from **ParentNode\<1\>**
- **n > 2** : I talked to @pet-mit about this case. He told me that, for expressions about **sum** node, constructor will need no more than 3 arguments, to be stored in 3 different well-named data members. So for this case, the **[] operator** is useless.
- **n = 2** : do we want to keep **[] operator** and **ParentNode\<n\>** only for this case ? We could have a left and a right data member, as it used to be. 

My suggestion : 
- We should drop classe **ParentNode\<n\>**, not useful.
- We can wonder whether we keep class **UnaryNode**, as it only allows to store one child node for many unary operators. Unfortunately, there is only one unary operator for now (**NegationNode**).
- Class **BinaryNode** should remain useful, but should not inherit from **ParentNode\<2\>**

a-zakir replied:
I don't agree with you, the intention behind BinaryNode and UnaryNode is clear, to represent nodes containing 1 or 2 nodes. The client of these classes knows exactly how to build them, and how to use them with the[] operator.

a-zakir replied:
_"n = 1 : using a [] operator is overkill. So any Node class that contains another Node should not inherit from ParentNode<1>"_ why would someone inherit from ParentNode<1> if he/she seeks to store more than one node?

flomnes replied:
I disagree as well. This allows to eliminate a lot of common code and avoid duplication.

> If it turns out that we really need ParentNode<n>, we can specialize its constructor, depending on the value of N.

No, this is no easy task (I tried and failed)

guilpier-code replied:
I meant : 
- if a class aims at containing just one other node (like **NegationNode**), the benefit for this class to inherit from **ParentNode<1>** is very small. Indeed, in general, the only benefit for a class to inherit from **ParentNode** is to reuse the **[] operator**. In this particular case, it's not a benefit : we always use the first element (index 0).
- So we may think that the benefit of reusing the **[] operator** of **ParentNode** could be higher if n > 1, right ? 
- But, as said above, n = 3 does not require the reuse of this operator either.
- Then if the class **ParentNode** is only used for case n = 2, its interest is limited, and this parent class is probably not useful. If you add to this the fact that it makes class hierarchy grow ("which is bad" some say), it can feel like removing it.

guilpier-code replied:
Was done in another PR. Let's close this chat.

pet-mit commented on src/solver/expressions/include/antares/solver/expressions/nodes/ParameterNode.h:
```diff
@@ -0,0 +1,14 @@
+
+namespace Antares::Solver::Nodes
+{
+class ParameterNode: public Leaf<std::string>
```
declare it as final?

a-zakir replied:
done

guilpier-code commented on src/solver/expressions/nodes/BinaryNode.cpp:
```diff
@@ -0,0 +1,31 @@
+/*
```
**Caution** : this file (and **UnaryNode.cpp**) is not referenced in any **cmake** file.
That said, if we follow [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2290/files?show-viewed-files=true&file-filters%5B%5D=#r1715177876), these files/classes is no more required

a-zakir replied:
i think your IDE is lying to you ;) [look](https://github.com/AntaresSimulatorTeam/Antares_Simulator/blob/bb55185b7b17150d429c3d04368832ff5c70d258/src/solver/expressions/CMakeLists.txt#L8)

guilpier-code replied:
ok, done.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/ComparisonNode.h:
```diff
@@ -0,0 +1,32 @@
+
+#include <antares/solver/expressions/nodes/BinaryNode.h>
+
+namespace Antares::Solver::Nodes
```
About class **ComparisonNode** : does not seem to be useful.

It has no data members, no behavior and is not visited (no visitor has an associated method).

flomnes replied:
Yes, it's present as a convenience class (ancestor to `GreaterThanOrEqualNode`, `LessThanOrEqualNode` and `EqualNode`). It can be removed just like `UnaryNode` and `BinaryNode` AFAIK.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/EqualNode.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace Antares::Solver::Nodes
+{
+class EqualNode: public ComparisonNode
```
As already said for **ComparisonNode.h**, not sure class **ComparisonNode** is useful

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,96 @@
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
```
Shouldn't the source file containing the definition of the parent of all visitors be located in the same directory than its children's ?
In short : in **src/solver/expressions/include/antares/solver/expressions/visitors**

flomnes replied:
No, the concrete visitors depend on abstract class `NodeVisitor<T>`.

a-zakir replied:
that was the original place for NodeVisitor, this class implement the dispatch [depending on node type ], hence belongs to nodes namespaces/lib/

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodesForwardDeclaration.h:
```diff
@@ -0,0 +1,42 @@
+#pragma once
+#include <string>
+
+namespace Antares::Solver::Nodes
```
Why do we need this header ?
Could not we use **ExpressionsNodes.h** instead ?

a-zakir replied:
to forward declare in headers; especially for visitors

flomnes replied:
We need this header for cases where only a declaration is required, not a definition. Think pointers, references, etc.

A good example is visitor headers
```cpp
class EvalVisitor: public Nodes::NodeVisitor<double>
{
    // ...
    double visit(const Nodes::AddNode& add) override;
    double visit(const Nodes::SubtractionNode& add) override;
    double visit(const Nodes::MultiplicationNode& add) override;
    double visit(const Nodes::DivisionNode& add) override;
    double visit(const Nodes::EqualNode& add) override;
    double visit(const Nodes::LessThanOrEqualNode& add) override;
    double visit(const Nodes::GreaterThanOrEqualNode& add) override;
    double visit(const Nodes::NegationNode& neg) override;
    double visit(const Nodes::VariableNode&) override;
    double visit(const Nodes::ParameterNode& param) override;
    double visit(const Nodes::LiteralNode& lit) override;
    double visit(const Nodes::PortFieldNode& port_field_node) override;
    double visit(const Nodes::ComponentVariableNode& component_variable_node) override;
    double visit(const Nodes::ComponentParameterNode& component_parameter_node) override;
};
```

guilpier-code replied:
ok. Let's close this chat.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NegationNode.h:
```diff
@@ -0,0 +1,34 @@
+
+namespace Antares::Solver::Nodes
+{
+class NegationNode: public UnaryNode
```
class **NegationNode** cannot be derived. We should add the **final** keyword.
More generally, we should add this keyword to all class not supposed to be derived.

a-zakir replied:
ok

a-zakir committed:
linear visitor (#2325)

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

guilpier-code commented on src/tests/src/solver/expressions/test_expressions.cpp:
```diff
@@ -0,0 +1,377 @@
+{
+    ParameterNode root("my-param");
+    const double value = 221.3;
+    EvaluationContext context({{"my-param", value}}, {});
```
About class **EvaluationContext** : 
To be honest, I don't understand the purpose of this class, I wonder why we need a context in the case of an evaluation, and don't need it in case of another visitor, and I wonder if it will eventually be useful or if it is there only for test purposes. In the latter case, it should be removed at some point.

From my current understanding, expressions are used to form constraints, and for that, parameters does not have to be evaluated : it seems we only use their name when building a constraint (ex : **level[t] - level[t-1] - efficiency * injection + withdrawal = inflows**).
So they don't really have a value. 
Do I miss something ?
If I'm right, unit tests about evaluating parameters or variables are not relevant in that form.
Unit tests should reflect the way we use object in real life. Here it's unclear how we'll use these concepts.

flomnes replied:
You are right, I think it's more likely that we'll have **partial** evaluation visitors. For example a visitor that transforms `ComponentParameter` nodes into `Literal` nodes through value substitution.

a-zakir committed:
update

a-zakir committed:
Merge branch 'feature/expression-visitors' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into feature/expression-visitors

a-zakir committed:
namespace

guilpier-code commented on src/solver/expressions/visitors/LinearVisitor.cpp:
```diff
@@ -0,0 +1,100 @@
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+
+#include "antares/solver/expressions/visitors/LinearVisitor.h"
```
**LinearVisitor.cpp** is not referenced in any **cmake** file.

a-zakir replied:
wrong

flomnes replied:
tests/src/solver/expressions/test_expressions.cpp has some

guilpier-code commented on src/solver/expressions/visitors/LinearVisitor.cpp:
```diff
@@ -0,0 +1,100 @@
+namespace Antares::Solver::Visitors
+{
+
+LinearStatus LinearVisitor::visit(const Nodes::AddNode& add)
```
In the following example : 
```c++
LinearStatus LinearVisitor::visit(const Nodes::AddNode& add)
{
    return dispatch(*add[0]) + dispatch(*add[1]);
}
```
how can we add 2 **LinearStatus** ?
==> ok :  **LinearStatus** has a **+ operator**

a-zakir replied:
Look at operators overloads in LinearStatus class

flomnes replied:
Yes that's the trick

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/visitors/LinearVisitor.h:
```diff
@@ -0,0 +1,51 @@
+
+class LinearStatus;
+
+class LinearVisitor: public Nodes::NodeVisitor<LinearStatus>
```
As this visitor evaluates the linearity of an expression, I would name it (for instance) **EvaluateLinearityVisitor**, or **LinearityEvaluationVisitor**, more clear in my opinion.

flomnes replied:
100% agree with you

a-zakir replied:
renamed LinearityVisitor

a-zakir committed:
add final keyword to Leaf nodes

a-zakir committed:
rename LinearityVisitor

a-zakir committed:
Pr comments

flomnes committed:
Move tryFunction out of class (#2332)

PROS
- Remove pointer to member functions

CONS
- Code looks a bit less readable ?

guilpier-code committed:
Expressions : remove class ParentNode (#2331)

This PR is a try to remove class **ParentNode** considered as useless in
[this
comment/discussion](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2290#discussion_r1715177876)

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,97 @@
+namespace
+{
+template<class RetT, class VisitorT, class NodeT>
+std::optional<RetT> tryType(const Node& node, VisitorT& visitor)
```
**tryType(...)** : this function has a fuzzy name, which does not tell much about what it does.
It actually tries to visit a node.
So it should be called **tryVisit**.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,97 @@
+namespace
+{
+template<class RetT, class VisitorT, class NodeT>
+std::optional<RetT> tryType(const Node& node, VisitorT& visitor)
```
**tryType(...)** : 
We just made this function a free function (see a previous PR).
While doing this, we added a new argument and a template parameter (type **VisitorT**) to it.
This function is now more complex : its definition and calls are more complex.

Why not make this function a member function of class **NodeVisitor** again ?
It would make sens : usually visitors have **visit** functions, I would not be shocked to have this function in the scope of class **NodeVisitor**.
Doing so, it would loose an argument and a template parameter.

flomnes replied:
Yes, there are pros and cons. The simplification is that we don't need pointer-to-member functions with their [complicated syntax](https://www.ibm.com/docs/en/i/7.4?topic=only-pointers-members-c).

a-zakir replied:
i'm resolving this conversation, @guilpier-code feel free to re-open the this thread

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,97 @@
+    R dispatch(const Node& node)
+    {
+        using FunctionT = std::optional<R> (*)(const Node&, NodeVisitor<R>&);
+        std::vector<FunctionT> functions{&tryType<R, NodeVisitor<R>, AddNode>,
```
**functions(...)** : for a collection of specific functions, we can do better than just name it **functions**.
 We could name it **allNodesVisitList** (for instance).

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,97 @@
+    R dispatch(const Node& node)
+    {
+        using FunctionT = std::optional<R> (*)(const Node&, NodeVisitor<R>&);
+        std::vector<FunctionT> functions{&tryType<R, NodeVisitor<R>, AddNode>,
```
**functions(...)** : to me, its a vector of pointer to functions having the same signature. 
Why not make it a vector of **std::function** ?

guilpier-code replied:
See already existing remark [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2290#discussion_r1716602665)

flomnes replied:
We don't need type erasure, stateful callable objects, etc. here. `std::function` adds a little bit of overhead (unnecessary in my opinion).

flomnes committed:
Move nodes/NodeVisitor.h -> visitors/NodeVisitor.h

flomnes committed:
PR remarks, auto*

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/expression-visitors

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h:
```diff
@@ -0,0 +1,97 @@
+    {
+        return visitor.visit(*x);
+    }
+    else
```
**Cosmetic** : removal of the **else** statement, and simply `return {};` after the **if** closing bracket.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/visitors/LinearityVisitor.h:
```diff
@@ -0,0 +1,51 @@
+class LinearityVisitor: public Nodes::NodeVisitor<LinearStatus>
+{
+public:
+    using Base = Nodes::NodeVisitor<LinearStatus>;
```
About `using Base = Nodes::NodeVisitor<LinearStatus>;` : 
I met this kind of alias in every visitor.
It seems it's not used.
If so, we should remove it.
If it turns out to be useful only internally, we should make it **private**, not **public**

a-zakir replied:
good point, the is a dead code

flomnes committed:
PR remarks

This reverts commit 50e28d256611082ee44dea6e711f449de443c03e.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/Registry.hxx:
```diff
@@ -0,0 +1,52 @@
+#include <mutex>
+#include <vector>
+
+namespace Antares::Solver
```
Why do we need a **.hxx** for that template class ? 
Would a simple **.h** not be convenient enough ?

a-zakir replied:
why not .hpp?

a-zakir replied:
To distinguish from template-free header (.h), we've choosen .hxx, but it could have been .hpp or something equivalent. I believe this project does not have convention on this subject

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/Registry.hxx:
```diff
@@ -0,0 +1,52 @@
+    {
+        auto created = std::make_unique<Derived>(std::forward<Args>(args)...);
+
+        std::lock_guard<std::mutex> lock(mutex_);
```
About `std::lock_guard<std::mutex> lock(mutex_);` :
Will we have to protect **registry_**  data member in a multi-threading context ?
I wonder : if we're in the context of a MC year (so inside a thread), maybe we should have a registry per year ?
Anyway, this protection is premature at this stage. We should introduce it if we really need it. For now, it makes code more complex than it should be.

a-zakir replied:
personally i can't rely on my memory ability, i prefer to do it once for all

flomnes replied:
Or we can use dependency injection [like this](https://gist.github.com/flomnes/54c89bc3ef48090bd3bf2e46692cc79d), but is it really worth it ?

a-zakir replied:
to replace one line of code? it definitely not worth it

guilpier-code replied:
In my opinion, there is very little chance that the registry (used in the context of expressions) is called from a thread, and gets in competition with other threads for creating a node.
Is it ?

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/Registry.hxx:
```diff
@@ -0,0 +1,52 @@
+namespace Antares::Solver
+{
+//  Template class to manage the memory allocation and registry for a base class
+template<class Base>
```
```C++
template<class Base>
class Registry
{
    ...
}
```
Do we need the `template<class Base>`, knowing that eventually the type will be **Node** ?
I guess it's in case we need to generalize the use of this **registry** concept.
But, for now, we don't need this generalization, so we should not implement it (see KISS principle : keep it simple, stupid).

a-zakir replied:
it could be, you may have noticed that this file is located outside visitors/nodes folders

flomnes replied:
Yes this is because we want to call virtual destructors to avoid memory leaks. If we didn't have a common ancestor, it wouldn't work.

guilpier-code replied:
What I recommend is an application of [this principe](https://www.geeksforgeeks.org/what-is-yagni-principle-you-arent-gonna-need-it/).

To be as clear as possible, I think we should have, instead : 
```C++
class Registry
{
public:
    ...
    template<class Derived, class... Args>
    requires std::derived_from<Derived, Nodes::Node>
    ...
private:
    std::vector<std::unique_ptr<Nodes::Node>> registry_; 
    ...
};
```

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/Registry.hxx:
```diff
@@ -0,0 +1,52 @@
+    //  Method to create a new derived class object and add it to the registry
+    template<class Derived, class... Args>
+    requires std::derived_from<Derived, Base>
+    Base* create(Args&&... args)
```
About **create** function : why not return the **Derived** type, that is the real type of the object we want to create ?

a-zakir replied:
Registry::registry_ is a vector of unique_ptr<Base> and even if the returned ptr is a derived ptr, it would be necessarily used as Base* in the context of visitor::dispatch method

flomnes replied:
Yes it's possible to return the derived type, I can make a PR if needed. It's slightly more convenient for the caller
### Current
```cpp
Node* add = registry.create<AddNode>(node1, node2);
Node* left = add->left(); // ERROR
Node* left = dynamic_cast<AddNode*>(add)->left(); // OK
```
### Proposed
```cpp
AddNode* add = registry.create<AddNode>(node1, node2); // Proposed
Node* left = add->left(); // OK
```

flomnes replied:
Done, d91ed8a8d

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/visitors/CloneVisitor.h:
```diff
@@ -0,0 +1,51 @@
+
+    Nodes::Node* visit(const Nodes::AddNode& add) override;
+    Nodes::Node* visit(const Nodes::SubtractionNode& add) override;
+    Nodes::Node* visit(const Nodes::MultiplicationNode& add) override;
```
`Nodes::Node* visit(const Nodes::MultiplicationNode& add)` : 
We should be consistent with arguments names : a **MultiplicationNode** should not be called **add**.
This little inconsistency occurs in many **visitors**.
In visitors' headers files, **visit(....)** functions don't need named argument by the way.

flomnes replied:
Yes in general I think it's a lot easier to just call arguments `node`.

flomnes replied:
All arguments are now named `node`

guilpier-code commented on src/solver/expressions/visitors/CloneVisitor.cpp:
```diff
@@ -0,0 +1,112 @@
+
+namespace Antares::Solver::Visitors
+{
+CloneVisitor::CloneVisitor(Registry<Nodes::Node>& mem):
```
`CloneVisitor(Registry<Nodes::Node>& mem)` : 
argument's name (**mem**) is not consistent with its type.
Let's make it simple :  `CloneVisitor(Registry<Nodes::Node>& registry)`

guilpier-code commented on src/solver/expressions/visitors/CloneVisitor.cpp:
```diff
@@ -0,0 +1,112 @@
+{
+}
+
+Nodes::Node* CloneVisitor::visit(const Nodes::AddNode& input_node)
```
`Nodes::Node* CloneVisitor::visit(const Nodes::AddNode& input_node)` : 
About the argument, input_node should be renamed into **addNode**, much closer to what it really is.

flomnes replied:
Or just `node`, see [comment above](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2290#discussion_r1723231767).

guilpier-code commented on src/solver/expressions/visitors/CloneVisitor.cpp:
```diff
@@ -0,0 +1,112 @@
+                                            dispatch(*input_node.right()));
+}
+
+Nodes::Node* CloneVisitor::visit(const Nodes::SubtractionNode& input_node)
```
Same as above  : input_node ==> **SubstractionNode**, and if we feel it's too long, **SubstNode**. But why not the first name (which is the clearest) ?

`LessThanOrEqualNode& input_node` ==> `LessThanOrEqualNode& lessNode` ?
`VariableNode& param` ==> `VariableNode& varNode` ?

flomnes replied:
Or just `node`, see [comment above](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2290#discussion_r1723231767).

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/Leaf.h:
```diff
@@ -0,0 +1,47 @@
+    {
+    }
+
+    T getValue() const
```
**T getValue() const** ==>` T value() const` ?
Shorter, clear enough, allows saving room at end of lines (when calling this function)

flomnes replied:
Yes, even `T operator() const` would do the job.

guilpier-code replied:
> Yes, even `T operator() const` would do the job.

It should remain clear, and **operator()** has no name.

flomnes replied:
Done, 1bb198ef0

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/Registry.hxx:
```diff
@@ -0,0 +1,52 @@
+{
+//  Template class to manage the memory allocation and registry for a base class
+template<class Base>
+class Registry
```
**Discussion**

Follows a talk with @flomnes.

This registry existence seems strange : it's both an object factory and a kind of garbage collector (it keeps memory of created object, in order to destroy them when it's destroyed). It kind of follows the RAII idiom.
It seems unnatural and we feel it could be more simple, and that we could make without it.

How could we get rid of this strange object ?

It seems we don't have clear benefits by removing the registry, and replacing it with usage of smart pointers on nodes.
- if nodes reference their possible child nodes with a **unique_ptr\<Node\>**, a child cannot be shared by another parent node
- if nodes reference their possible child nodes with a **shared_ptr\<Node\>**, we will have an overhead at execution, due to references counter management. 

We'll see if registry makes the job with time.
 
Please feel free to add any useful comment.

flomnes replied:
This memory register offers some flexibility, and is relatively simple to use. 

On the other hand `std::shared_ptr` incur some overhead, so maybe we should test it for large trees.

a-zakir committed:
draft time index visitor

a-zakir committed:
Revert "draft time index visitor"

This reverts commit 0c703cbc60c6aa3ceaad08290053f76657cd7005.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/visitors/LinearStatus.h:
```diff
@@ -0,0 +1,140 @@
+namespace Antares::Solver::Visitors
+{
+
+class LinearStatus
```
**_Draft remark - needs to be completed_**

This class contains code with quite high complexity, with **if... else...** statements nested inside **switch** statement. This is also error prone if this code has to evolve.
We also have a class **LinearStatus** containing just one data members and many operators (+, -, *, /).

I was wondering if, in order to simplify this, we could have something like : 

```c++
enum class LinearStatus
{
    CONSTANT,
    LINEAR,
    NON_LINEAR
};

LinearStatus operator+ ()

```
_to be finished / corrected..._

flomnes replied:
See #2343

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/visitors/NodeVisitor.h:
```diff
@@ -0,0 +1,95 @@
+public:
+    virtual ~NodeVisitor() = default;
+
+    R dispatch(const Node& node)
```
**dispatch(...)** : what's the idea behind this name ?
**dispatch** mostly means **send**.
In a classic visitor pattern, no one "sends" anything.
Instead, **visit** would be a candidate, or **visitNode**. 
**broadcast** ? **traverse** ? If we mean that we iterate over a tree.

flomnes replied:
**dispatch** the sending of someone or something to a destination or for a purpose.

I think it's a good name, clearer than the propositions.

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/EqualNode.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace Antares::Solver::Nodes
+{
+class EqualNode: public ComparisonNode
```
class **EqualNode** : 
Looking at function `PrintVisitor::visit(const EqualNode& equ)`, this class represents the comparison '=='.
Is it possible to have a '==' in an expression ?
Sign '=' can also be met in expressions, and does not exist as a **Node**.

If I miss something, sorry. But if I don't, there might be a confusion between, '=' and '=='.
And in that case, why do we have : 
```C++
LinearStatus LinearityVisitor::visit(const Nodes::EqualNode& equ)
{
    return LinearStatus::NON_LINEAR;
}
```
?
The non linearity of this '=' operator will cause the whole expression to be non linear. 
I would not expect an expression containing a '=' to be necessarily non linear.

flomnes replied:
This operator is meant to represent constraints or equations, such as `3x+4y=2`.

Indeed, I think that `x=y` is linear if and only if x and y are linear.

flomnes replied:
See #2351

flomnes committed:
Add comparison visitor, using variadic template arguments (#2342)

- Add variadic argument to `NodeVisitor::visit` methods. This argument
is optional, meaning we don't break existing visitors
- Add `CompareVisitor` a non-commutative comparison visitor to another
expression (a+b != b+a)
- Add simple & complex tests, both in the case where expressions are
expected to be equal, and different.
- Add comparison operators (`operator==`) for some nodes, but not all.
Open to remarks

flomnes committed:
Use enum class & free functions for `LinearStatus`, add unit tests (#2343)

Some functions may have changed behavior, add some unit tests to avoid regressions.

flomnes committed:
Substitution from context visitor (#2333)

Co-authored-by: Abdoulbari Zaher <32519851+a-zakir@users.noreply.github.com>

flomnes committed:
PR remarks

flomnes committed:
PR remarks

guilpier-code commented on src/solver/expressions/visitors/PrintVisitor.cpp:
```diff
@@ -0,0 +1,99 @@
+    return "(" + dispatch(*node.left()) + "/" + dispatch(*node.right()) + ")";
+}
+
+std::string PrintVisitor::visit(const Nodes::EqualNode& node)
```
`std::string PrintVisitor::visit(const Nodes::EqualNode& node)` : 
"==" does not seem to exist in expressions.
Shouldn't we have "=" instead ?

guilpier-code commented on src/tests/src/solver/expressions/test_expressions.cpp:
```diff
@@ -0,0 +1,579 @@
+    VariableNode var2("y");
+    // x==y
+    Node* eq = create<EqualNode>(&var1, &var2);
+    BOOST_CHECK_EQUAL(printVisitor.dispatch(*eq), "x==y");
```
`BOOST_CHECK_EQUAL(printVisitor.dispatch(*eq), "x==y");` : 
Is "==" supposed to be part of a expression ?
Otherwise, it should be replaced with "="

a-zakir committed:
enum case (#2349)

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

a-zakir committed:
Split expression-lib tests  (#2352)

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Fix LinearVisitor behavior for comparison nodes, add tests (#2351)

flomnes committed:
Add tests with slightly larger trees (#2353)

flomnes committed:
Registry::create returns Derived*

flomnes committed:
Rename getValue -> value

pet-mit commented on src/solver/expressions/include/antares/solver/expressions/nodes/ComponentNode.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace Antares::Solver::Nodes
+{
+class ComponentNode: public Node
```
why will we need this class in production code?

flomnes replied:
I's only required for ComponentVariable & ComponentParameter, to avoid code duplication.

We might keep it in a private header instead.

pet-mit replied:
oh ok thanks I did not see the heritage!

flomnes committed:
Remove unused headers (#2355)

flomnes committed:
Increase test coverage for TimeIndexVisitor (#2357)

flomnes committed:
Fix operator for boost data (#2356)

`bdata::make(...) ^ bdata::make(...)` zips datasets 

{a1, a2} ^ {b1, b2} = {{a1, b1}, {a2, b2}}


https://www.boost.org/doc/libs/1_85_0/boost/test/data/monomorphic/zip.hpp

`bdata::make(...) * bdata::make(...)` makes a grid of datasets (all
(x,y) pairs)

{a1, a2} * {b1, b2} = {{a1, b1}, {a1, b2}, {a2, b1}, {a2, b2}}



https://www.boost.org/doc/libs/1_85_0/boost/test/data/monomorphic/grid.hpp

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/ComponentNode.h:
```diff
@@ -0,0 +1,53 @@
+    const std::string& getComponentId() const;
+    const std::string& getComponentName() const;
+
+    bool operator==(const ComponentNode& other) const;
```
Do we need the definition of **operator ==** ? 
I mean : since **C++20**, we can have : 
```C++
bool operator==(const ComponentNode& other) const == default;
```

flomnes replied:
TL;DR : no, since an ancestor class has `deleted` `operator==`

https://gist.github.com/flomnes/d39b3b59a415edadb78c0b0be7994a3b

flomnes replied:
EDIT : done #2359

guilpier-code commented on src/solver/expressions/nodes/PortFieldNode.cpp:
```diff
@@ -0,0 +1,45 @@
+{
+}
+
+bool PortFieldNode::operator==(const PortFieldNode& other) const
```
Remark already made for other node classes : since C++20, we can have **default** operator==

flomnes replied:
(see answer above)

guilpier-code commented on src/tests/src/solver/expressions/test_CompareVisitor.cpp:
```diff
@@ -0,0 +1,123 @@
+#include <antares/solver/expressions/nodes/ExpressionsNodes.h>
+#include <antares/solver/expressions/visitors/CloneVisitor.h>
+#include <antares/solver/expressions/visitors/CompareVisitor.h>
+#include <antares/solver/expressions/visitors/PrintVisitor.h>
```
```C++
#include <antares/solver/expressions/visitors/PrintVisitor.h>
```
It seems we don't need **PrintVisitor** here.

guilpier-code commented on src/tests/src/solver/expressions/test_CompareVisitor.cpp:
```diff
@@ -0,0 +1,123 @@
+using namespace Antares::Solver::Nodes;
+using namespace Antares::Solver::Visitors;
+
+// Only necessary for BOOST_CHECK_EQUAL
```
```C++
// Only necessary for BOOST_CHECK_EQUAL
```
comment needed ?

flomnes replied:
no

guilpier-code commented on src/tests/src/solver/expressions/test_CompareVisitor.cpp:
```diff
@@ -0,0 +1,123 @@
+    return expr;
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_self_simple, Registry<Node>)
```
Renaming : **comparison_to_self_simple** ==> **compare_a_simple_node_to_itself**
Easier to read when unit test title is a short sentence.

guilpier-code commented on src/tests/src/solver/expressions/test_CompareVisitor.cpp:
```diff
@@ -0,0 +1,123 @@
+    BOOST_CHECK(cmp.dispatch(*expr, *expr));
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_other_same, Registry<Node>)
```
Renaming : **comparison_to_other_same** ==> **compare_a_simple_node_to_an_identical_node**

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/visitors/CompareVisitor.h:
```diff
@@ -0,0 +1,51 @@
+    bool visit(const Nodes::AddNode& add, const Nodes::Node& other) override;
+    bool visit(const Nodes::SubtractionNode& add, const Nodes::Node& other) override;
+    bool visit(const Nodes::MultiplicationNode& add, const Nodes::Node& other) override;
+    bool visit(const Nodes::DivisionNode& add, const Nodes::Node& other) override;
```
Renaming : 
- **add** ==> **node** 
- **neg** ==> **node**
- ...

guilpier-code commented on src/solver/expressions/visitors/CompareVisitor.cpp:
```diff
@@ -0,0 +1,132 @@
+}
+
+template<class T>
+static bool compareGetValue(const T& node, const Antares::Solver::Nodes::Node& other)
```
Renaming : **compareGetValue** ==> **compareValue**

guilpier-code commented on src/tests/src/solver/expressions/test_CompareVisitor.cpp:
```diff
@@ -0,0 +1,123 @@
+    BOOST_CHECK(!cmp.dispatch(*expr1, *expr2));
+}
+
+static Node* createComplexExpression(Registry<Node>& registry)
```
We first test a comparison of very simple expressions, then very complex ones.
Shouldn't we test a comparison of half-complex ones ?

flomnes committed:
Use default operator== (#2359)

guilpier-code commented on src/solver/expressions/visitors/CompareVisitor.cpp:
```diff
@@ -0,0 +1,132 @@
+#include <antares/solver/expressions/visitors/CompareVisitor.h>
+
+template<class T, class V>
+static bool compareBinaryNode(V& visitor, const T& node, const Antares::Solver::Nodes::Node& other)
```
About helper free function **compareBinaryNode** : 
I would **not** be shocked if this function was a **private method** of class **CompareVisitor**.
It seems natural, and it would simplify it (the visitor argument may disappear, as well as a template parameter) : 
```C++
template<class T>
bool CompareVisitor::compareBinaryNode(const T& node, 
                                       const Nodes::Node& other)
{
    if (const T* other_node = dynamic_cast<const T*>(&other))
    {
        bool left = dispatch(*node.left(), *other_node->left());
        bool right = dispatch(*node.right(), *other_node->right());
        return left && right;
    }
    return false;
}
```
Would it make sense ?

a-zakir replied:
the function does not change any object [class] state;

guilpier-code replied:
The would have less arguments, and its body would be bit simpler.

guilpier-code commented on src/solver/expressions/visitors/CompareVisitor.cpp:
```diff
@@ -0,0 +1,132 @@
+#include <antares/solver/expressions/visitors/CompareVisitor.h>
+
+template<class T, class V>
+static bool compareBinaryNode(V& visitor, const T& node, const Antares::Solver::Nodes::Node& other)
```
Let's be crazy and try to remove the remaining template parameter : 
Could we have something like this : 
```C++
bool CompareVisitor::operator==(const Nodes::BinaryNode* n1, 
                                const Nodes::Node* n2)
{
    if (typeid(*n1) == typeid(*n2))
    {
        return dispatch(n1->left(), n2->left()) && 
               dispatch(n1->right(), n2->right());
    }
    return false;
}
```
?

guilpier-code commented on src/solver/expressions/visitors/CompareVisitor.cpp:
```diff
@@ -0,0 +1,132 @@
+}
+
+template<class T>
+static bool compareEqualOperator(const T& node, const Antares::Solver::Nodes::Node& other)
```
**compareEqualOperator** : 
if suggestions on **compareBinaryNode** work (not guaranteed) and are suitable, we could try them on **compareEqualOperator** and **compareGetValue**

guilpier-code commented on src/solver/expressions/include/antares/solver/expressions/nodes/Node.h:
```diff
@@ -0,0 +1,31 @@
+{
+public:
+    virtual ~Node() = default;
+    bool operator==(const Node& other) const = default;
```
```C++
bool operator==(const Node& other) const = default;
```
Is this useful, for itself (as it does not have attributes) or its children ?

guilpier-code commented on src/tests/src/solver/expressions/test_main.cpp:
```diff
@@ -0,0 +1,25 @@
+ * along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+ */
+
+#define BOOST_TEST_MODULE expressions
```
Is this file useful ? For what ?

a-zakir replied:
to produce 1 exe for multiple translation units

a-zakir replied:
... 1 for each test suite

guilpier-code commented on src/tests/src/solver/expressions/test_CompareVisitor.cpp:
```diff
@@ -0,0 +1,123 @@
+    BOOST_CHECK(cmp.dispatch(*expr, *expr));
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_clone_complex, Registry<Node>)
```
Here is something I don't understand : I suppose that, in real life, we should associate one registry to one expression / tree. And it would make sense : a clone life time should be independent from the life time of the tree it was created from.
Here we have one registry for a tree and its clone. Is it intended ?

guilpier-code replied:
If the previous remark is relevant, something else can be pointed : 
if we must associate one registry to one tree / expression, can we have 2 independent objects alive for an expression (I mean the registry and the expression itself) ?
User of an expression must handle 2 objects, which are mutually cohesive : one cannot exist without the other.
So if we think that registry class is useful, we should find a way the 'link' more closely expressions and registry. For example having a class **Expression** or **Tree** or **AST** that contain a root node and a registry as data member.

We still have the solution where node are pointed to with **std::shared_ptr** (which we know to be flawed).

a-zakir replied:
the notion of tree is not defined here, we only care about nodes.

guilpier-code replied:
I rephrase my question : 
here we have only have one registry for 2 expressions (or tree) : **expr** and **expr_cloned**.
Is it normal ?
Should not we have 2 registries : one associated to each expressions ?

guilpier-code commented on src/tests/src/solver/expressions/test_CompareVisitor.cpp:
```diff
@@ -0,0 +1,123 @@
+    BOOST_CHECK(cmp.dispatch(*expr, *expr));
+}
+
+BOOST_FIXTURE_TEST_CASE(comparison_to_clone_complex, Registry<Node>)
```
If the previous remark is relevant, something else can be pointed : 
if we must associate one registry to one tree / expression, can we have 2 independent objects alive for an expression (I mean the registry and the expression itself) ?
User of an expression must handle 2 objects, which are mutually cohesive : one cannot exist without the other.
So if we think that registry class is useful, we should find a way the 'link' more closely an expression and its associated registry. For example by having a class **Expression** or **Tree** or **AST** that contain a root node and a registry as data members.

We still have the solution where nodes are pointed to with **std::shared_ptr** at creation (which we know to be flawed).

a-zakir replied:
there is nothing that forbids to use multiple registries or none at all!

guilpier-code commented on src/tests/src/solver/expressions/test_TimeIndexVisitor.cpp:
```diff
@@ -0,0 +1,193 @@
+} // namespace Antares::Solver::Visitors
+BOOST_AUTO_TEST_SUITE(_TimeIndexVisitor_)
+
+BOOST_FIXTURE_TEST_CASE(simple_time_dependant_expression, Registry<Node>)
```
It is recommended not to make unit tests that long : 
it makes many things, so it could be more clear, it could be easier to identify why it fails, ... 
It's explained [there](https://brightsec.com/blog/unit-testing-best-practices/)  (for example).

guilpier-code commented on src/tests/src/solver/expressions/test_TimeIndexVisitor.cpp:
```diff
@@ -0,0 +1,193 @@
+    PrintVisitor printVisitor;
+    std::unordered_map<const Node*, TimeIndex> context;
+    // LiteralNode --> constant in time and for all scenarios
+    LiteralNode literalNode(65.);
```
Renaming : 
`LiteralNode literalNode(65.);` ==> `LiteralNode literal(65.);` (shorter, without losing in expressivity), or `LiteralNode sixtyFive(65.);`

guilpier-code commented on src/tests/src/solver/expressions/test_TimeIndexVisitor.cpp:
```diff
@@ -0,0 +1,193 @@
+{
+    switch (s)
+    {
+    case TimeIndex::CONSTANT_IN_TIME_AND_SCENARIO:
```
Renaming : `TimeIndex::CONSTANT_IN_TIME_AND_SCENARIO` ==> `TimeIndex::CONSTANT`

a-zakir replied:
was changed from TimeIndex::CONSTANT -->TimeIndex::CONSTANT_IN_TIME_AND_SCENARIO

guilpier-code commented on src/tests/src/solver/expressions/test_TimeIndexVisitor.cpp:
```diff
@@ -0,0 +1,193 @@
+    {
+    case TimeIndex::CONSTANT_IN_TIME_AND_SCENARIO:
+        return os << "TimeIndex::CONSTANT_IN_TIME_AND_SCENARIO";
+    case TimeIndex::VARYING_IN_TIME_ONLY:
```
Renaming suggestions : 
`TimeIndex::VARYING_IN_TIME_ONLY` ==> `TimeIndex::TIME_DEPENDENT` (word **time-dependent** is very common in scientific english)

`VARYING_IN_SCENARIO_ONLY` ==> `SCENARIO_DEPENDENT`

a-zakir replied:
i think the word "only" give more information about the time/scenario

a-zakir committed:
doxygen doc for expression lib (#2361)

flomnes committed:
Simplify recursion for deepAddTree function (#2366)

a-zakir committed:
optimize trytype loop (#2362)

The current dispatch method iterates through a list of functions
**trytype**, sequentially checking each one to see if it can handle the
given node type. This can be inefficient.
in this PR we use unordered_map as look up table between node type and
trytype function:
less comparisons && faster access than sequential iteration

---------

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

guilpier-code committed:
Test improvement trial (#2358)

This is a suggestion for a unit tests enhancement.
Unit tests are a bit simpler like this.
The main idea is to introduce a **fixture** containing more things than
just a **registry**.
What do you think ?

a-zakir committed:
use raw ptr instead of ref (#2360)

* control nullity of input node (will be done in #2354 )
* avoid de-referencing Nodes::Node* -> visitor(Node&) (null ref can be
obtained from null ptr )
* avoid indirection: visitor was working with Node& and Nodes::Node
childs are Node* (Node& --> Node*)

a-zakir committed:
pr

a-zakir committed:
sonar issues from #2290 (#2354)
